# Real-world Data Wrangling

In this project, you will apply the skills you acquired in the course to gather and wrangle real-world data with two datasets of your choice.

You will retrieve and extract the data, assess the data programmatically and visually, accross elements of data quality and structure, and implement a cleaning strategy for the data. You will then store the updated data into your selected database/data store, combine the data, and answer a research question with the datasets.

Throughout the process, you are expected to:

1. Explain your decisions towards methods used for gathering, assessing, cleaning, storing, and answering the research question
2. Write code comments so your code is more readable

Before you start, install the some of the required packages. 

# Research Question 

### How are housing price and Homelessness attributes correlated? 


## Definitions


df = This is the first data, from kaggle that Have been scrapped from Realtor website. 

df2 = This is the second data, The Homelessness, in the period 2007 - 2016 - Kaggle. 


In [74]:
!python -m pip install kaggle==1.6.12

### **1.1.** Problem Statement
In 2-4 sentences, explain the kind of problem you want to look at and the datasets you will be wrangling for this project.

>I would like to work on Real estate in USA

In [6]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
import requests
import zipfile
import json
import requests
import os



## 1. Gather data

In this section, you will extract data using two different data gathering methods and combine the data. Use at least two different types of data-gathering methods.

### **1.2.** Gather at least two datasets using two different data gathering methods

## Dataset 1 
The first data set is a spreadsheet a full of informations about the USA housing data, prices, places and dates and of sold or not, I downlowded it manully from Kaggle, and this data was collected from https://www.realtor.com/ - A real estate listing website operated by the News Corp subsidiary Move, Inc. and based in Santa Clara, California. 
I choosed this dataset to study the housing market in USA. Download it manually. 


In [15]:
df = pd.read_csv("realtor-data.zip.csv")


## Dataset 2

The Seconed Dataset is about homelessness is USA it has 2 csv files, homlessness in the period 2007 - 2016 and population csv. 
These raw data sets contain Point-in-Time (PIT) estimates and national PIT estimates of homelessness as well as national estimates of homelessness by state and estimates of chronic homelessness from 2007 - 2016. Estimates of homeless veterans are also included beginning in 2011. The accompanying Housing Inventory Count (HIC) data is available as well from 2007 - 2016." (Department of Housing and Urban Development).
> Type: zip File, CSV File

> Method: The data was gathered using the "API" method from Kaggle.
> Dataset variables:
Index(['Year', 'State', 'CoC Number', 'CoC Name', 'Measures', count

omes in $1000's

In [26]:
import kaggle

In [28]:
from kaggle.api.kaggle_api_extended import KaggleApi


In [69]:

zip_file_path = "homelessness.zip"  # Replace with the path to the downloaded zip file

# Define the directory to extract the files to
extract_dir = "2007-2016-Homelessnewss-USA.csv"  # Replace with the directory where you want to extract the files

# Unzip the downloaded file
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(extract_dir)

# Remove the downloaded zip file
os.remove(zip_file_path)






FileNotFoundError: [Errno 2] No such file or directory: '2007-2016-Homelessnewss-USA.csv\\2007-2016-Homelessnewss-USA.csv'

In [71]:
df2 = pd.read_csv('2007-2016-Homelessnewss-USA.csv/2007-2016-Homelessnewss-USA.csv')

FileNotFoundError: [Errno 2] No such file or directory: '2007-2016-Homelessnewss-USA.csv/2007-2016-Homelessnewss-USA.csv'

In [46]:
df.head()

brokered_by    status     price  bed  bath  acre_lot     street  \
0     103378.0  for_sale  105000.0  3.0   2.0      0.12  1962661.0   
1      52707.0  for_sale   80000.0  4.0   2.0      0.08  1902874.0   
2     103379.0  for_sale   67000.0  2.0   1.0      0.15  1404990.0   
3      31239.0  for_sale  145000.0  4.0   2.0      0.10  1947675.0   
4      34632.0  for_sale   65000.0  6.0   2.0      0.05   331151.0   

         city        state  zip_code  house_size prev_sold_date  
0    Adjuntas  Puerto Rico     601.0       920.0            NaN  
1    Adjuntas  Puerto Rico     601.0      1527.0            NaN  
2  Juana Diaz  Puerto Rico     795.0       748.0            NaN  
3       Ponce  Puerto Rico     731.0      1800.0            NaN  
4    Mayaguez  Puerto Rico     680.0         NaN            NaN

In [48]:
df.tail(10)

brokered_by status     price  bed  bath  acre_lot     street  \
2226372     108243.0   sold  425000.0  3.0   3.0      0.06   970797.0   
2226373      16235.0   sold  305000.0  4.0   2.0      0.42   353937.0   
2226374      53860.0   sold  310000.0  3.0   1.0      0.21   500240.0   
2226375      60631.0   sold  385000.0  4.0   2.0      0.21   210890.0   
2226376      85499.0   sold  339900.0  4.0   2.0      0.20    41160.0   
2226377      23009.0   sold  359900.0  4.0   2.0      0.33   353094.0   
2226378      18208.0   sold  350000.0  3.0   2.0      0.10  1062149.0   
2226379      76856.0   sold  440000.0  6.0   3.0      0.50   405677.0   
2226380      53618.0   sold  179900.0  2.0   1.0      0.09   761379.0   
2226381     108243.0   sold  580000.0  5.0   3.0      0.31   307704.0   

             city       state  zip_code  house_size prev_sold_date  
2226372  Richland  Washington   99354.0      1876.0     2022-02-14  
2226373  Richland  Washington   99354.0      2000.0     2022-02-11  
2226374  Richland  Washington   99354.0      1152.0     2022-02-11  
2226375  Richland  Washington   99354.0      1656.0     2022-03-28  
2226376  Richland  Washington   99354.0      2780.0     2022-03-28  
2226377  Richland  Washington   99354.0      3600.0     2022-03-25  
2226378  Richland  Washington   99354.0      1616.0     2022-03-25  
2226379  Richland  Washington   99354.0      3200.0     2022-03-24  
2226380  Richland  Washington   99354.0       933.0     2022-03-24  
2226381  Richland  Washington   99354.0      3615.0     2022-03-23

In [50]:
df.describe() 

brokered_by         price           bed          bath      acre_lot  \
count  2.221849e+06  2.224841e+06  1.745065e+06  1.714611e+06  1.900793e+06   
mean   5.293989e+04  5.241955e+05  3.275841e+00  2.496440e+00  1.522303e+01   
std    3.064275e+04  2.138893e+06  1.567274e+00  1.652573e+00  7.628238e+02   
min    0.000000e+00  0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    2.386100e+04  1.650000e+05  3.000000e+00  2.000000e+00  1.500000e-01   
50%    5.288400e+04  3.250000e+05  3.000000e+00  2.000000e+00  2.600000e-01   
75%    7.918300e+04  5.500000e+05  4.000000e+00  3.000000e+00  9.800000e-01   
max    1.101420e+05  2.147484e+09  4.730000e+02  8.300000e+02  1.000000e+05   

             street      zip_code    house_size  
count  2.215516e+06  2.226083e+06  1.657898e+06  
mean   1.012325e+06  5.218668e+04  2.714471e+03  
std    5.837635e+05  2.895408e+04  8.081635e+05  
min    0.000000e+00  0.000000e+00  4.000000e+00  
25%    5.063128e+05  2.961700e+04  1.300000e+03  
50%    1.012766e+06  4.838200e+04  1.760000e+03  
75%    1.521173e+06  7.807000e+04  2.413000e+03  
max    2.001357e+06  9.999900e+04  1.040400e+09

In [52]:
df.columns

Index(['brokered_by', 'status', 'price', 'bed', 'bath', 'acre_lot', 'street',
       'city', 'state', 'zip_code', 'house_size', 'prev_sold_date'],
      dtype='object')

In [54]:
!pip install kaggle

In [56]:
print(df.isnull().sum())


brokered_by         4533
status                 0
price               1541
bed               481317
bath              511771
acre_lot          325589
street             10866
city                1407
state                  8
zip_code             299
house_size        568484
prev_sold_date    734297
dtype: int64


In [58]:
print(df.duplicated().sum())


0


In [65]:
api = KaggleApi()
api.authenticate()
api.dataset_download_files("adamschroeder/homelessness" ,
                           "2007-2016-Homelessnewss-USA.csv")

Dataset URL: https://www.kaggle.com/datasets/adamschroeder/homelessness


FileNotFoundError: [Errno 2] No such file or directory: '2007-2016-Homelessnewss-USA.csv\\homelessness.zip'

In [86]:
#FILL IN 1st data gathering and loading method

## 2. Assess data

Assess the data according to data quality and tidiness metrics using the report below.

List **two** data quality issues and **two** tidiness issues. Assess each data issue visually **and** programmatically, then briefly describe the issue you find.  **Make sure you include justifications for the methods you use for the assessment.**

"dirty data and messy data.

Inaccurate data is an example of dirty data, i.e., a data quality issue. High data quality is crucial for obtaining reliable insights and driving informed decision-makin.

Unorganized data is an example of messy data, i.e., a data structural issue. With tidy data, we can focus on performing additional assessments and cleaning without dealing with roadblocks around getting the data into a structure that's easier to view and pa" 
 data.rse.

In [67]:
df2.head(10)

NameError: name 'df2' is not defined

In [218]:
df2.tail(10)

Year State CoC Number               CoC Name  \
86519  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86520  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86521  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86522  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86523  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86524  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86525  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86526  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86527  1/1/2016    WY     WY-500  Wyoming Statewide CoC   
86528  1/1/2016    WY     WY-500  Wyoming Statewide CoC   

                                                Measures Count  
86519                               Unsheltered Homeless   366  
86520                   Unsheltered Homeless Individuals   240  
86521            Unsheltered Homeless People in Families   126  
86522  Unsheltered Homeless Unaccompanied Children (U...     2  
86523  Unsheltered Homeless Unaccompanied Young Adult...     5  
86524  Unsheltered Homeless Unaccompanied Youth (Unde...     7  
86525                      Unsheltered Homeless Veterans    31  
86526             Unsheltered Parenting Youth (Under 25)     3  
86527              Unsheltered Parenting Youth Age 18-24     3  
86528               Unsheltered Parenting Youth Under 18     0

In [222]:
df2.describe()

Year  State CoC Number       CoC Name  \
count      86529  86529      86529          86529   
unique        10     54        414            414   
top     1/1/2015     CA     AK-500  Anchorage CoC   
freq       16926   8946        216            216   

                                Measures  Count  
count                              86529  86529  
unique                                42   3608  
top     Chronically Homeless Individuals      0  
freq                                3999  12209

In [224]:
print(df2.isnull().sum())


Year          0
State         0
CoC Number    0
CoC Name      0
Measures      0
Count         0
dtype: int64


In [226]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Year        86529 non-null  object
 1   State       86529 non-null  object
 2   CoC Number  86529 non-null  object
 3   CoC Name    86529 non-null  object
 4   Measures    86529 non-null  object
 5   Count       86529 non-null  object
dtypes: object(6)
memory usage: 4.0+ MB


In [230]:
df2.columns

Index(['Year', 'State', 'CoC Number', 'CoC Name', 'Measures', 'Count'], dtype='object')

In [236]:
df2.value_counts()

Year      State  CoC Number  CoC Name                                 Measures                                           Count
1/1/2007  AK     AK-500      Anchorage CoC                            Chronically Homeless Individuals                   224      1
1/1/2015  IL     IL-504      Madison County CoC                       Homeless Unaccompanied Youth (Under 25)            1        1
                                                                      Sheltered Chronically Homeless People in Families  10       1
                                                                      Sheltered Chronically Homeless Individuals         5        1
                                                                      Sheltered Chronically Homeless                     15       1
                                                                                                                                 ..
1/1/2012  CO     CO-503      Metropolitan Denver Homeless Initiative  Chronically

In [ ]:
#Optional: store the raw data in your local data store


> data quality:

  * *Completeness: is a metric that helps you understand whether your data is sufficient to answer interesting questions or solve your problem*

 *  *Validity is a metric helping you understand how well your data conforms to a defined set of rules for data, also known as a schema* 
   
* Accuracy is a metric that helps you understand whether your data accurately represents the reality it aims to depict.* 
* Consistency is a metric that helps you understand two things: whether your data follows a standard format and whether your data’s info matches with information from other data sources.
* Uniqueness is a metric that helps you understand whether there are duplicate or overlapping values in your data.

### Quality Issue and Tidiness Issues (Notes for myself)

#FILL IN - Inspecting the data frame visually 

*  The shape of the df is huge It covers a lot of years, but df2 covers from 2007 to 2016: *So We have to limit the shape according to year. 
**( drop the data after 2017 | 12/2016  so it fits with the second dataset, or Filter the DataFrame to include only data from 2007 to 2016)** 
> And the code for this is: 
 df = df[(df['Year'] >= 2007) & (df['Year'] <= 2016)]   
- Year here is (prev_sold_date) So it have to be renamed either ways to ["Year"] which is object have to be changed to datetime.
- The prev_sold_date  object, and it must be datatime. And it's important to notice that this column gives us an information saying this the "Sold date", So I have to use this information correctly to connect the 2 datasets, it's important to notice it might be a key column instead of the state.
-  I have in df['prev_sold_date'] a 734297 null value.. first, I will reshape the data, then decide on how to deal with 
    date null values. 

* In df2 "count" must be integers.

* df Have a lot of nulls. However, the most needed ones is the prices and the state. And df2 does not have any nulls.
* To Fix this issue, I must define the necessary columns in both datas. 
  - necessary_columnsdf = df['Price', 'State', 'sold_date/ Year']
  - necessary_columnsdf2 = df2['State', 'Year', 'Measures', 'counts']
* * df have 8 missing Values in ['State'] I have to change it to ['unknown'] or fill it manually by knowing the city. 

* In df2 ["States"] must be defined by name So I have to change it be matches df1.

  > df2['State'] = df2['State'].replace({'Alaska':'AK', 'Alabama':'AL', 'Arkansas':'AR', 'Arizona':'AZ', 
                      'California':'CA', 'Colorado':'CO', 'Connecticut':'CT', 
                      'District of Columbia':'DC', 'Delaware':'DE', 'Florida':'FL', 
                      'Georgia':'GA', 'Hawaii':'HI', 'Iowa':'IA', 
                      'Idaho':'ID', 'Illinois':'IL', 'Indiana':'IN', 'Kansas':'KS', 
                      'Kentucky':'KY', 'Louisiana':'LA', 'Massachusetts':'MA', 'Maryland':'MD', 
                      'Maine':'ME', 'Michigan':'MI', 'Minnesota':'MN', 'Missouri':'MO', 
                      'Mississippi':'MS', 'Montana':'MT', 'North Carolina':'NC', 
                      'North Dakota':'ND', 'Nebraska':'NE', 'New Hampshire':'NH', 
                      'New Jersey':'NJ', 'New Mexico':'NM', 'Nevada':'NV', 'New York':'NY', 
                      'Ohio':'OH', 'Oklahoma':'OK', 'Oregon':'OR', 'Pennsylvania':'PA', 
                      'Puerto Rico':'PR', 'Rhode Island':'RI', 'South Carolina':'SC', 
                      'South Dakota':'SD', 'Tennessee':'TN', 'Texas':'TX', 'Utah':'UT', 
                      'Virginia':'VA', 'Vermont':'VT', 'Washington':'WA', 
                      'Wisconsin':'WI', 'West Virginia':'WV', 'Wyoming':'WY'})

* In df2 Measures addresses many kind of "Homelessness", which are Homeless Individuals',' Homeless People in Families', 'Unsheltered Homeless', We can make it moinge Tidy make it i \\\\\ also there is " Unsheltered Parenting Youth Under 18" And "...Under 25.." n one category
   > hodf2ols = homeless['Measures'].str.df2ies()
homeless.join(homeless_cols[['Homeless Individuals','Homeless People in Families', 'Unshelter
     - Homeless']]*
* But still,I dont think we have to make it in categories.-  Instead, I Think I have to collect the "count" number in one re
  ult per state. 
 bject

In [270]:
df.shape

(2226382, 12)

In [272]:
df2.shape

(86529, 6)

In [283]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2226382 entries, 0 to 2226381
Data columns (total 12 columns):
 #   Column          Dtype  
---  ------          -----  
 0   brokered_by     float64
 1   status          object 
 2   price           float64
 3   bed             float64
 4   bath            float64
 5   acre_lot        float64
 6   street          float64
 7   city            object 
 8   state           object 
 9   zip_code        float64
 10  house_size      float64
 11  prev_sold_date  object 
dtypes: float64(8), object(4)
memory usage: 203.8+ MB


In [285]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Year        86529 non-null  object
 1   State       86529 non-null  object
 2   CoC Number  86529 non-null  object
 3   CoC Name    86529 non-null  object
 4   Measures    86529 non-null  object
 5   Count       86529 non-null  object
dtypes: object(6)
memory usage: 4.0+ MB


In [309]:
df.sample(5)

brokered_by    status     price  bed  bath  acre_lot     street  \
1940016      31129.0      sold  199900.0  2.0   2.0      0.09   659198.0   
1517756       1213.0      sold  235000.0  3.0   2.0      0.17   412293.0   
1646114      29175.0      sold  169900.0  4.0   2.0      0.40  1416373.0   
120563       24196.0  for_sale  250000.0  NaN   1.0       NaN   848947.0   
255690       71656.0  for_sale  399900.0  3.0   2.0       NaN   223001.0   

                   city         state  zip_code  house_size prev_sold_date  
1940016       Arlington         Texas   76013.0      1468.0     2021-12-16  
1517756       Lancaster  Pennsylvania   17603.0      1460.0     2022-03-09  
1646114     Forest Park       Georgia   30297.0      1753.0     2022-02-24  
120563    New York City      New York   11201.0       420.0            NaN  
255690   Virginia Beach      Virginia   23453.0      2193.0     2021-07-16

In [311]:
df.columns

Index(['brokered_by', 'status', 'price', 'bed', 'bath', 'acre_lot', 'street',
       'city', 'state', 'zip_code', 'house_size', 'prev_sold_date'],
      dtype='object')

In [276]:
df2.sample(5)

Year State CoC Number  \
83944  1/1/2016    SC     SC-502   
21835  1/1/2011    IL     IL-519   
1185   1/1/2007    GA     GA-503   
165    1/1/2007    AR     AR-504   
1935   1/1/2007    MA     MA-511   

                                                CoC Name  \
83944                              Columbia/Midlands CoC   
21835                          West Central Illinois CoC   
1185                            Athens/Clarke County CoC   
165                                      Delta Hills CoC   
1935   Quincy/Brockton/Weymouth/Plymouth City and Cou...   

                                         Measures Count  
83944       Sheltered Homeless People in Families   252  
21835  Sheltered Chronically Homeless Individuals     9  
1185                         Unsheltered Homeless   131  
165                          Unsheltered Homeless   510  
1935   Sheltered Chronically Homeless Individuals    61

In [278]:
print(df.isnull().sum())

brokered_by         4533
status                 0
price               1541
bed               481317
bath              511771
acre_lot          325589
street             10866
city                1407
state                  8
zip_code             299
house_size        568484
prev_sold_date    734297
dtype: int64


In [306]:
print(df.columns)

Index(['brokered_by', 'status', 'price', 'bed', 'bath', 'acre_lot', 'street',
       'city', 'state', 'zip_code', 'house_size', 'prev_sold_date'],
      dtype='object')


In [280]:
print(df2.isnull().sum())

Year          0
State         0
CoC Number    0
CoC Name      0
Measures      0
Count         0
dtype: int64


In [655]:
df2['Measures'].unique()

array(['Chronically Homeless Individuals', 'Homeless Individuals',
       'Homeless People in Families',
       'Sheltered Chronically Homeless Individuals', 'Sheltered Homeless',
       'Sheltered Homeless Individuals',
       'Sheltered Homeless People in Families', 'Total Homeless',
       'Unsheltered Chronically Homeless Individuals',
       'Unsheltered Homeless', 'Unsheltered Homeless Individuals',
       'Unsheltered Homeless People in Families', 'Chronically Homeless',
       'Chronically Homeless People in Families', 'Homeless Veterans',
       'Sheltered Chronically Homeless',
       'Sheltered Chronically Homeless People in Families',
       'Sheltered Homeless Veterans', 'Unsheltered Chronically Homeless',
       'Unsheltered Chronically Homeless People in Families',
       'Unsheltered Homeless Veterans', 'Children of Parenting Youth',
       'Homeless Unaccompanied Children (Under 18)',
       'Homeless Unaccompanied Young Adults (Age 18-24)',
       'Homeless Unaccompan

In [746]:
df2['Measures'].nunique()

42

In [752]:
df2['Measures']. duplicated().sum()

85839

In [756]:
df2['Measures']. isnull().sum()

0

In [671]:
df2['Measures'].value_counts

<bound method IndexOpsMixin.value_counts of 0                         Chronically Homeless Individuals
1                                     Homeless Individuals
2                              Homeless People in Families
3               Sheltered Chronically Homeless Individuals
4                                       Sheltered Homeless
                               ...                        
86524    Unsheltered Homeless Unaccompanied Youth (Unde...
86525                        Unsheltered Homeless Veterans
86526               Unsheltered Parenting Youth (Under 25)
86527                Unsheltered Parenting Youth Age 18-24
86528                 Unsheltered Parenting Youth Under 18
Name: Measures, Length: 86529, dtype: object>

## 3. Clean data
Clean the data to solve the 4 issues corresponding to data quality and tidiness found in the assessing step. **Make sure you include justifications for your cleaning decisions.**

After the cleaning for each issue, please use **either** the visually or programatical method to validate the cleaning was succesful.

At this stage, you are also expected to remove variables that are unnecessary for your analysis and combine your datasets. Depending on your datasets, you may choose to perform variable combination and elimination before or after the cleaning stage. Your dataset must have **at least** 4 variables after combining the data.

# For the porpuse of combine 

Here is to make the time zone for df the same as df2: 
- Rename prev_sold_date To "Year". 
- dtype: datetime
- 1/1/2006 - 31/12/2017 

In [318]:
df.rename(columns= {'prev_sold_date': 'Year'}, inplace= True)  

Check: 

In [380]:
df.sample(1)

brokered_by    status     price  bed  bath  acre_lot    street  \
1305875      85342.0  for_sale  434900.0  3.0   2.0      0.16  363868.0   

                city       state  zip_code  house_size       Year  
1305875  San Jacinto  California   92583.0      1290.0 2007-11-21

Change the dtype for Year because it's object

In [353]:
df['Year'] = pd.to_datetime(df['Year'])

Check

In [356]:
df.dtypes

brokered_by           float64
status                 object
price                 float64
bed                   float64
bath                  float64
acre_lot              float64
street                float64
city                   object
state                  object
zip_code              float64
house_size            float64
Year           datetime64[ns]
dtype: object

Now, lemit the df dataframe

In [358]:
df = df[(df['Year'].dt.year >= 2007) & (df['Year'].dt.year <= 2016)]


Check, The shape of df was (2226382, 12)

In [361]:
df.shape

(199269, 12)

In [366]:
check = 2226382 - 199569
print(check)

2026813


So, I've dropped 2026813 intries. I have to check the nulls. 

In [383]:
print(df.isnull().sum())

brokered_by      151
status             0
price             69
bed            21374
bath           20529
acre_lot       29659
street           828
city              18
state              0
zip_code           1
house_size     30312
Year               0
dtype: int64


Now,
- I have no nulls in the Year Column. 
- I have no nulls in the State Column
- I have 69 in the Price column. I can fill it with th mean of every year to care about the accuracy. 

### **Quality Issue 1: FILL IN**

In [414]:
# FILL IN - Validate the cleaning was successful
# Here,  I must fill the Na in prices with the mean of the corresponding year.

# Calculate the mean price for each year
mean_prices = df.groupby('Year')['price'].transform('mean')

# Fill missing prices with the mean price of the corresponding year
df['price'] = df['price'].fillna(mean_prices)

# Display the DataFrame after filling missing prices
print("DataFrame after filling missing prices with the mean price of each year:")
print(df)


DataFrame after filling missing prices with the mean price of each year:
         brokered_by    status      price  bed  bath  acre_lot     street  \
2270         94933.0  for_sale   950000.0  5.0   4.0      0.99  1260473.0   
2493         22968.0  for_sale    79500.0  NaN   NaN      0.76   765975.0   
2555         34543.0  for_sale   176500.0  NaN   NaN      0.50   917409.0   
3062         22968.0  for_sale   129000.0  NaN   NaN      6.30   739696.0   
3409         21163.0  for_sale   525000.0  3.0   3.0      0.45  1813270.0   
...              ...       ...        ...  ...   ...       ...        ...   
1414189      79471.0  for_sale   425000.0  3.0   2.0      0.16  1728797.0   
1414192      95190.0  for_sale   414900.0  4.0   2.0      0.23  1706555.0   
1414316      21688.0  for_sale   525000.0  3.0   2.0      0.46  1003563.0   
1414327      57432.0  for_sale  1100000.0  4.0   4.0      0.32  1280711.0   
1414346      82020.0  for_sale   497000.0  4.0   3.0      0.41  1698210.0   

  

In [426]:
print(df['price'])

2270        950000.0
2493         79500.0
2555        176500.0
3062        129000.0
3409        525000.0
             ...    
1414189     425000.0
1414192     414900.0
1414316     525000.0
1414327    1100000.0
1414346     497000.0
Name: price, Length: 199269, dtype: float64


Check: 

In [422]:
print(df.isnull().sum())

brokered_by      151
status             0
price              0
bed            21374
bath           20529
acre_lot       29659
street           828
city              18
state              0
zip_code           1
house_size     30312
Year               0
dtype: int64


Justification: *FILL IN*

### **Quality Issue 2: FILL IN**

In [439]:
df2.columns

Index(['Year', 'State', 'CoC Number', 'CoC Name', 'Measures', 'Count'], dtype='object')

In [449]:
#FILL IN - Apply the cleaning strategy
# In df2 "count" must be integers
df2['Count'] = df2['Count'].str.replace(',', '')

df2['Count'] = df2['Count'].astype(int)

Check

In [452]:
#FILL IN - Validate the cleaning was successful
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Year        86529 non-null  object
 1   State       86529 non-null  object
 2   CoC Number  86529 non-null  object
 3   CoC Name    86529 non-null  object
 4   Measures    86529 non-null  object
 5   Count       86529 non-null  int32 
dtypes: int32(1), object(5)
memory usage: 3.6+ MB


Now, The Years Column is object here too, and it must be datetime too. 

In [463]:
df2['Year'] = pd.to_datetime(df2['Year'])

In [465]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86529 entries, 0 to 86528
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Year        86529 non-null  datetime64[ns]
 1   State       86529 non-null  object        
 2   CoC Number  86529 non-null  object        
 3   CoC Name    86529 non-null  object        
 4   Measures    86529 non-null  object        
 5   Count       86529 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(4)
memory usage: 3.6+ MB


Justification: *FILL IN*

### **Tidiness Issue 1: FILL IN**

In [618]:
#FILL IN - Apply the cleaning strategy
# Here I have to change the state column so when I combine the data make sense. 

state_abbr_to_full = {
    'AK': 'Alaska', 'AL': 'Alabama', 'AR': 'Arkansas', 'AZ': 'Arizona', 'CA': 'California',
    'CO': 'Colorado', 'CT': 'Connecticut', 'DC': 'District of Columbia', 'DE': 'Delaware', 'FL': 'Florida',
    'GA': 'Georgia', 'HI': 'Hawaii', 'IA': 'Iowa', 'ID': 'Idaho', 'IL': 'Illinois', 'IN': 'Indiana',
    'KS': 'Kansas', 'KY': 'Kentucky', 'LA': 'Louisiana', 'MA': 'Massachusetts', 'MD': 'Maryland',
    'ME': 'Maine', 'MI': 'Michigan', 'MN': 'Minnesota', 'MO': 'Missouri', 'MS': 'Mississippi', 'MT': 'Montana',
    'NC': 'North Carolina', 'ND': 'North Dakota', 'NE': 'Nebraska', 'NH': 'New Hampshire', 'NJ': 'New Jersey',
    'NM': 'New Mexico', 'NV': 'Nevada', 'NY': 'New York', 'OH': 'Ohio', 'OK': 'Oklahoma', 'OR': 'Oregon',
    'PA': 'Pennsylvania', 'PR': 'Puerto Rico', 'RI': 'Rhode Island', 'SC': 'South Carolina', 'SD': 'South Dakota',
    'TN': 'Tennessee', 'TX': 'Texas', 'UT': 'Utah', 'VA': 'Virginia', 'VT': 'Vermont', 'WA': 'Washington',
    'WI': 'Wisconsin', 'WV': 'West Virginia', 'WY': 'Wyoming', 'VI': 'Virgin Islands', 'GU': 'Guam'
}

# Replace state abbreviations with full names
df2['State'] = df2['State'].replace(state_abbr_to_full)

#States in df2 but not in df:
#{'Puerto Rico', 'VI', 'GU'}

#States in df but not in df2:
#{'Virgin Islands'}

In [620]:
print(df['State'].unique())

['Virgin Islands' 'Massachusetts' 'Connecticut' 'New Hampshire' 'Vermont'
 'New York' 'Rhode Island' 'Maine' 'New Jersey' 'Pennsylvania' 'Delaware'
 'Ohio' 'West Virginia' 'Maryland' 'Virginia' 'District of Columbia'
 'North Carolina' 'Kentucky' 'Tennessee' 'South Carolina' 'Georgia'
 'Alabama' 'Florida' 'Mississippi' 'Arkansas' 'Missouri' 'Louisiana'
 'Indiana' 'Illinois' 'Michigan' 'Wisconsin' 'Iowa' 'Minnesota'
 'South Dakota' 'Nebraska' 'North Dakota' 'Montana' 'Idaho' 'Kansas'
 'Oklahoma' 'Colorado' 'Wyoming' 'Texas' 'New Mexico' 'Utah' 'Nevada'
 'Washington' 'Oregon' 'Arizona' 'California' 'Hawaii' 'Alaska']


In [622]:
print(df2['State'].unique())

['Alaska' 'Alabama' 'Arkansas' 'Arizona' 'California' 'Colorado'
 'Connecticut' 'District of Columbia' 'Delaware' 'Florida' 'Georgia'
 'Guam' 'Hawaii' 'Iowa' 'Idaho' 'Illinois' 'Indiana' 'Kansas' 'Kentucky'
 'Louisiana' 'Massachusetts' 'Maryland' 'Maine' 'Michigan' 'Minnesota'
 'Missouri' 'Mississippi' 'Montana' 'North Carolina' 'North Dakota'
 'Nebraska' 'New Hampshire' 'New Jersey' 'New Mexico' 'Nevada' 'New York'
 'Ohio' 'Oklahoma' 'Oregon' 'Pennsylvania' 'Puerto Rico' 'Rhode Island'
 'South Carolina' 'South Dakota' 'Tennessee' 'Texas' 'Utah' 'Virginia'
 'Virgin Islands' 'Vermont' 'Washington' 'Wisconsin' 'West Virginia'
 'Wyoming']


In [504]:
df2.sample(5)

Year         State CoC Number  \
83387 2016-01-01  Pennsylvania     PA-509   
37557 2013-01-01       Florida     FL-511   
81212 2016-01-01      New York     NY-508   
49968 2014-01-01      New York     NY-503   
73115 2016-01-01       Florida     FL-512   

                                                CoC Name  \
83387                           Eastern Pennsylvania CoC   
37557           Pensacola/Escambia/Santa Rosa County CoC   
81212  Buffalo, Niagara Falls/Erie, Niagara, Orleans,...   
49968                           Albany City & County CoC   
73115                             Saint Johns County CoC   

                                      Measures  Count  
83387  Homeless Unaccompanied Youth (Under 25)    209  
37557                     Chronically Homeless    243  
81212                       Sheltered Homeless    905  
49968                     Chronically Homeless     81  
73115       Sheltered Parenting Youth Under 18      0

In [570]:
df2['State'].nunique()

54

In [574]:
df['state']. nunique()

52

In [534]:
df.duplicated(). sum()

0

In [536]:
df2.duplicated(). sum()

0

I have 2 more issues 
> 
1- The state is in df2 in capital and in df is small. Must change to capital 
2- There is 2 more "State" in df2. 

In [580]:
df.rename(columns={'state': 'State'}, inplace=True)


In [584]:
df.head()

brokered_by    status     price  bed  bath  acre_lot     street  \
2270      94933.0  for_sale  950000.0  5.0   4.0      0.99  1260473.0   
2493      22968.0  for_sale   79500.0  NaN   NaN      0.76   765975.0   
2555      34543.0  for_sale  176500.0  NaN   NaN      0.50   917409.0   
3062      22968.0  for_sale  129000.0  NaN   NaN      6.30   739696.0   
3409      21163.0  for_sale  525000.0  3.0   3.0      0.45  1813270.0   

               city           State  zip_code  house_size       Year  
2270   Saint Thomas  Virgin Islands     802.0      5000.0 2013-10-11  
2493  Christiansted  Virgin Islands     820.0         NaN 2016-04-28  
2555  Christiansted  Virgin Islands     820.0         NaN 2016-11-16  
3062   Frederiksted  Virgin Islands     840.0         NaN 2010-01-26  
3409         Agawam   Massachusetts    1001.0      2314.0 2014-06-25

In [624]:

# Get the unique states in both DataFrames
unique_states_df = set(df['State'].unique())
unique_states_df2 = set(df2['State'].unique())

# Find states that are in df2 but not in df
states_in_df2_not_in_df = unique_states_df2 - unique_states_df

# Find states that are in df but not in df2
states_in_df_not_in_df2 = unique_states_df - unique_states_df2

print("States in df2 but not in df:")
print(states_in_df2_not_in_df)

print("\nStates in df but not in df2:")
print(states_in_df_not_in_df2)




States in df2 but not in df:
{'Puerto Rico', 'Guam'}

States in df but not in df2:
set()


#### So, I must drop States in df2 but not in df: 
{'Puerto Rico', 'Guam' 

- Now, I want to check the year column for tidiness purposes. I have noticed that the entries of the Year column are not tidy in df but they are tidy in df2.}

In [631]:
df =df.sort_values(by='Year').reset_index()

In [649]:
df.head(5)

index  brokered_by    status     price  bed  bath  acre_lot     street  \
0   911844      15977.0  for_sale   80000.0  NaN   NaN      3.00   659959.0   
1   864685       8748.0  for_sale  541000.0  4.0   3.0      0.25   811025.0   
2   829710      22490.0  for_sale  259000.0  2.0   3.0      0.25    77733.0   
3   864687       8748.0  for_sale  491245.0  4.0   3.0      0.25   811057.0   
4  1138769      44642.0  for_sale  410000.0  NaN   NaN      0.11  1253561.0   

           city      State  zip_code  house_size       Year  
0        Mexico   Missouri   65265.0         NaN 2007-01-01  
1    Plainfield   Illinois   60586.0      3084.0 2007-01-01  
2  Grand Rapids  Minnesota   55744.0      2360.0 2007-01-01  
3    Plainfield   Illinois   60586.0      2612.0 2007-01-01  
4      Superior   Colorado   80027.0         NaN 2007-01-02

In [646]:
df2.sample(5)

Year          State CoC Number  \
19013 2010-01-01     Washington     WA-507   
76825 2016-01-01       Maryland     MD-501   
47416 2014-01-01      Louisiana     LA-507   
72292 2016-01-01       Colorado     CO-500   
69573 2015-01-01  West Virginia     WV-508   

                                 CoC Name  \
19013            Yakima City & County CoC   
76825                  Baltimore City CoC   
47416    Alexandria/Central Louisiana CoC   
72292       Colorado Balance of State CoC   
69573  West Virginia Balance of State CoC   

                                    Measures  Count  
19013         Sheltered Homeless Individuals    194  
76825  Unsheltered Parenting Youth Age 18-24      0  
47416         Sheltered Homeless Individuals     98  
72292                   Chronically Homeless    703  
69573               Parenting Youth Under 18      0

# Measures 
*I have 42 kinds about Homeless people, So I will make them in 2 categories*

1- Sheltered 
2- Unsheltered.
However, I Still will have ['Chronically Homeless'] and ['Total Homeless'] not recognized. I will consider "Chronically Homeless" as Unsheltered Homeless Because "Chronically" must mean that. 
And, Total homeless means both sheltered and unsheltered I will consider the number in both sheltered and unsheltered. 


In [777]:
# Define keywords for categorization
sheltered_keywords = ['Sheltered', 'Parenting Youth', 'Chronically Homeless', 'Total Homeless', 'Homeless People in Families'] 
unsheltered_keywords = ['Unsheltered', 'Chronically Homeless', 'Homeless Unaccompanied', 'Total Homeless', 'Homeless Individuals']

# Categorize measures into 'Sheltered' and 'Unsheltered'
df2.loc[:, 'Sheltered'] = df2['Measures'].apply(lambda x: any(keyword in x for keyword in sheltered_keywords))
df2.loc[:, 'Unsheltered'] = df2['Measures'].apply(lambda x: any(keyword in x for keyword in unsheltered_keywords))


# Display DataFrame
print(df2)

         State       Year                                           Measures  \
0       Alaska 2007-01-01                   Chronically Homeless Individuals   
1       Alaska 2007-01-01                               Homeless Individuals   
2       Alaska 2007-01-01                        Homeless People in Families   
3       Alaska 2007-01-01         Sheltered Chronically Homeless Individuals   
4       Alaska 2007-01-01                                 Sheltered Homeless   
...        ...        ...                                                ...   
86524  Wyoming 2016-01-01  Unsheltered Homeless Unaccompanied Youth (Unde...   
86525  Wyoming 2016-01-01                      Unsheltered Homeless Veterans   
86526  Wyoming 2016-01-01             Unsheltered Parenting Youth (Under 25)   
86527  Wyoming 2016-01-01              Unsheltered Parenting Youth Age 18-24   
86528  Wyoming 2016-01-01               Unsheltered Parenting Youth Under 18   

       Count  Sheltered  Unsheltered  


In [773]:
df2.columns

Index(['State', 'Year', 'Measures', 'Count', 'Sheltered', 'Unsheltered'], dtype='object')

I have To drop Measures, After I make a copy.
**Drop the 'Measures' column**
df2 = df2.drop(columns=['Measures'])

### **Remove unnecessary variables and combine datasets**

Depending on the datasets, you can also peform the combination before the cleaning steps.

In [695]:
#FILL IN - Remove unnecessary variables and combine datasets
# necessary_columnsdf = df['Price', 'State', 'Year']
# necessary_columnsdf2 = df2['State', 'Year', 'Measures', 'counts']

necessary_df = df[['price','State', 'Year']]
necessary_df2 = df2 [['State', 'Year', 'Measures', 'Count']]

df = necessary_df
df2 = necessary_df2



Check

In [706]:
df.head(2)

price     State       Year
0   80000.0  Missouri 2007-01-01
1  541000.0  Illinois 2007-01-01

In [708]:
df2.head(2)

State       Year                          Measures  Count
0  Alaska 2007-01-01  Chronically Homeless Individuals    224
1  Alaska 2007-01-01              Homeless Individuals    696

In [758]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85881 entries, 0 to 86528
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   State     85881 non-null  object        
 1   Year      85881 non-null  datetime64[ns]
 2   Measures  85881 non-null  object        
 3   Count     85881 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(2)
memory usage: 2.9+ MB


In [760]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199269 entries, 0 to 199268
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   price   199269 non-null  float64       
 1   State   199269 non-null  object        
 2   Year    199269 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 4.6+ MB


- So, I must drop States in df2 but not in df: 
{'Puerto Rico', 'Guam}'

In [732]:
# States in df but not in df2
states_to_drop = {'Puerto Rico', 'Guam'}

# Drop rows in df2 where the state is in states_to_drop
df2 = df2[~df2['State'].isin(states_to_drop)]

In [734]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85881 entries, 0 to 86528
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   State     85881 non-null  object        
 1   Year      85881 non-null  datetime64[ns]
 2   Measures  85881 non-null  object        
 3   Count     85881 non-null  int32         
dtypes: datetime64[ns](1), int32(1), object(2)
memory usage: 2.9+ MB


## 4. Update your data store
Update your local database/data store with the cleaned data, following best practices for storing your cleaned data:

- Must maintain different instances / versions of data (raw and cleaned data)
- Must name the dataset files informatively
- Ensure both the raw and cleaned data is saved to your database/data store

# DF 2 - Homelessness in USA

In [789]:
# Save raw data to CSV
df2.to_csv('raw_homeless_data.csv', index=False)

In [793]:
df2 = df2.drop(columns=['Measures']) #drop the Measures data

KeyError: "['Measures'] not found in axis"

In [795]:
df2.head()

State       Year  Count  Sheltered  Unsheltered
0  Alaska 2007-01-01    224       True         True
1  Alaska 2007-01-01    696      False         True
2  Alaska 2007-01-01    278       True        False
3  Alaska 2007-01-01    187       True         True
4  Alaska 2007-01-01    842       True        False

In [797]:
df2.to_csv('cleaned_homeless_data.csv', index=False)

print("Raw and cleaned data have been saved to CSV files.")


Raw and cleaned data have been saved to CSV files.


# DF 1 : Real Estate in USA 

In [803]:
df.to_csv('cleaned_Houses_MArket_USA_data.csv', index=False)

In [805]:
df.head(2)

price     State       Year
0   80000.0  Missouri 2007-01-01
1  541000.0  Illinois 2007-01-01

In [817]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199269 entries, 0 to 199268
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   price   199269 non-null  float64       
 1   State   199269 non-null  object        
 2   Year    199269 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 4.6+ MB


In [839]:
df.loc[:, 'price'] = df['price'].astype(np.float32)

In [843]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199269 entries, 0 to 199268
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   price   199269 non-null  float32       
 1   State   199269 non-null  object        
 2   Year    199269 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float32(1), object(1)
memory usage: 3.8+ MB


In [841]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85881 entries, 0 to 86528
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   State        85881 non-null  object        
 1   Year         85881 non-null  datetime64[ns]
 2   Count        85881 non-null  int32         
 3   Sheltered    85881 non-null  bool          
 4   Unsheltered  85881 non-null  bool          
dtypes: bool(2), datetime64[ns](1), int32(1), object(1)
memory usage: 2.5+ MB


In [871]:
df2.loc[:, 'Count'] = df2['Count'].astype(np.int16)

In [873]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85881 entries, 0 to 86528
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   State        85881 non-null  object        
 1   Year         85881 non-null  datetime64[ns]
 2   Count        85881 non-null  int32         
 3   Sheltered    85881 non-null  bool          
 4   Unsheltered  85881 non-null  bool          
dtypes: bool(2), datetime64[ns](1), int32(1), object(1)
memory usage: 2.5+ MB


# Combine 2 datasets

In [8]:
combined_df = pd.merge(df, df2, on='State', how='inner')

NameError: name 'df' is not defined

## 5. Answer the research question

### **5.1:** Define and answer the research question 
Going back to the problem statement in step 1, use the cleaned data to answer the question you raised. Produce **at least** two visualizations using the cleaned data and explain how they help you answer the question.

*Research question:* FILL IN from answer to Step 1

In [ ]:
#Visual 1 - FILL IN

*Answer to research question:* FILL IN

In [ ]:
#Visual 2 - FILL IN

*Answer to research question:* FILL IN

### **5.2:** Reflection
In 2-4 sentences, if you had more time to complete the project, what actions would you take? For example, which data quality and structural issues would you look into further, and what research questions would you further explore?

*Answer:* FILL IN